client/state/data-layer/wpcom-http/utils.js

```diff
@@ -92,7 +108,11 @@ export const dispatchRequest = ( initiator, onSuccess, onError, options ) => ( s
 
 	const data = getData( action );
 	if ( data ) {
-		return onSuccess( store, action, data );
+		const [ isValid, response ] = fromApi( data );
+
+		return isValid
+			? onSuccess( store, action, response )
+			: onError( store, action, response );
```

Comment:
Interesting thoughts here @dmsnell. I'm wondering: 

If the server returns an error due to, for instance, a missing parameter and includes a message to be rendered to the UI via an `onError` function, the `response` here is the result of a failed validation attempt.
```js
[ false, validator.errors ]
```
`onError` would not have access to the message from the server, correct? What would be the best way accomodate this use case?

Reply:
strange, I remember responding here but it must have failed to get through

@psealock I don't think this is really an issue though I'm open to explore it. I think there are two scenarios where your concerns are raised and so I'll respond to what I understand:

**failure status code, failure data**

if a response comes back with a non 2xx status code it will be siphoned off into the `onError` handler before checking the validation of the response, so this shouldn't be a real issue. the error handler will get the original data from the server

**success status code, failure date**

some of our APIs return a status code in contradiction to the information contained in the message. if we need to respond to this then we still need to validate the shape of the response. that is, if we don't know what to expect we can't deal with it. in this case we may end up having JSON schema defining both success and error shapes for the success response

```json
{
   "$schema" : "http://json-schema.org/draft-04/schema#",
   "id" : "apiResponse",
   "title" : "apiResponse",
   "oneOf" : [
      {
         "additionalProperties" : false,
         "properties" : {
               "value" : {
                  "type": "integer",
               }
         },
         "type": "object"
      }, {
         "additionalProperties" : false,
         "properties" : {
               "error": {
                  "type": "string"
               }
            },
         },
         "type" : "object"
      }
   ]
}
```

something like that

Reply:
Gotcha, makes sense.